src/main/java/org/prebid/server/bidder/richaudience/RichaudienceBidder.java

```diff
@@ -177,24 +209,34 @@ private List<BidderBid> extractBids(BidResponse bidResponse, BidRequest bidReque
             return Collections.emptyList();
         }
 
+        final Map<String, Imp> impMap = bidRequest.getImp().stream()
+                .collect(Collectors.toMap(Imp::getId, Function.identity()));
+
         return bidResponse.getSeatbid().stream()
                 .filter(Objects::nonNull)
-                .map(SeatBid::getBid)
+                .flatMap(seatBid -> seatBid.getBid().stream())
                 .filter(Objects::nonNull)
-                .flatMap(Collection::stream)
-                .map(bid -> BidderBid.of(bid,
-                        resolvedBidType(bid.getImpid(), bidRequest.getImp()), bidResponse.getCur()))
+                //todo: is it fine to return a bid with unknown bid type? (as the PBS Go does)
+                .filter(bid -> impMap.containsKey(bid.getImpid()))
+                .map(bid -> makeBid(bidResponse, impMap, bid))
                 .toList();
     }
 
-    private static BidType resolvedBidType(String impId, List<Imp> imps) {
-        for (Imp imp : imps) {
-            if (impId.equals(imp.getId())) {
-                if (imp.getVideo() != null) {
-                    return BidType.video;
-                }
-                return BidType.banner;
-            }
+    private static BidderBid makeBid(BidResponse bidResponse, Map<String, Imp> impMap, Bid bid) {
+        final Imp imp = impMap.get(bid.getImpid());
+        final BidType bidType = resolveBidType(imp);
+        final Bid.BidBuilder builder = bid.toBuilder();
+
+        if (bidType == BidType.video) {
+            builder.w(imp.getVideo().getW());
+            builder.h(imp.getVideo().getH());
```

Comment:
Let's make some defensive programming things)) and make same checks if imp.getVideo() is null

Reply:
agree